File: packages/recorder_macos/macos/recorder_macos/core/Sources/RecorderCore/ResourceCensus.swift
import Foundation

/// What the host is still holding, counted (§19.1).
///
/// §19.1 observes that "state is cleaned up" cannot be a requirement, because
/// nothing can fail it. This is what makes it falsifiable: every row is an
/// integer, and the two tests it names — census equality across ten start →
/// stop cycles, and every row of its first table zero after one stop — are both
/// equality assertions on this.
///
/// It lives out here rather than in the plugin so `swift test` can reach the
/// arithmetic, and so the three things that answer it — the session, the
/// overlay windows and the meter — each report their own rows and are summed
/// once, instead of one of them owning a map the other two write into.
///
/// The wire shape is `[String: Int]`, matching `ResourceCensus` in
/// `recorder_platform_interface` key for key
/// (`docs/architecture/platform-channel-contract.md`).
public struct ResourceCensus: Equatable {
  public var captureStreams: Int
  public var cameraSessions: Int
  public var microphoneSessions: Int
  public var meteringTaps: Int
  public var meterSubscriptions: Int
  public var registeredTextures: Int
  public var overlayEngines: Int
  public var eventMonitors: Int
  public var sessionTimers: Int
  public var powerAssertions: Int
  public var writers: Int
  public var compositors: Int

  public init(
    captureStreams: Int = 0,
    cameraSessions: Int = 0,
    microphoneSessions: Int = 0,
    meteringTaps: Int = 0,
    meterSubscriptions: Int = 0,
    registeredTextures: Int = 0,
    overlayEngines: Int = 0,
    eventMonitors: Int = 0,
    sessionTimers: Int = 0,
    powerAssertions: Int = 0,
    writers: Int = 0,
    compositors: Int = 0
  ) {
    self.captureStreams = captureStreams
    self.cameraSessions = cameraSessions
    self.microphoneSessions = microphoneSessions
    self.meteringTaps = meteringTaps
    self.meterSubscriptions = meterSubscriptions
    self.registeredTextures = registeredTextures
    self.overlayEngines = overlayEngines
    self.eventMonitors = eventMonitors
    self.sessionTimers = sessionTimers
    self.powerAssertions = powerAssertions
    self.writers = writers
    self.compositors = compositors
  }

  /// Row-wise addition. Each contributor counts only what it owns, so the sum
  /// is the whole census and nothing is counted twice.
  public static func + (lhs: ResourceCensus, rhs: ResourceCensus) -> ResourceCensus {
    ResourceCensus(
      captureStreams: lhs.captureStreams + rhs.captureStreams,
      cameraSessions: lhs.cameraSessions + rhs.cameraSessions,
      microphoneSessions: lhs.microphoneSessions + rhs.microphoneSessions,
      meteringTaps: lhs.meteringTaps + rhs.meteringTaps,
      meterSubscriptions: lhs.meterSubscriptions + rhs.meterSubscriptions,
      registeredTextures: lhs.registeredTextures + rhs.registeredTextures,
      overlayEngines: lhs.overlayEngines + rhs.overlayEngines,
      eventMonitors: lhs.eventMonitors + rhs.eventMonitors,
      sessionTimers: lhs.sessionTimers + rhs.sessionTimers,
      powerAssertions: lhs.powerAssertions + rhs.powerAssertions,
      writers: lhs.writers + rhs.writers,
      compositors: lhs.compositors + rhs.compositors)
  }

  /// True when every row of §19.1's **first** table is zero.
  ///
  /// `overlayEngines` is excluded deliberately: §19.1's second table allows a
  /// host to keep its overlay engines for the life of the process, which is
  /// what this one does. What that host owes instead is stability, and the
  /// census *equality* test is what holds it to that.
  public var sessionResourcesReleased: Bool {
    captureStreams == 0 && cameraSessions == 0 && microphoneSessions == 0
      && meteringTaps == 0 && meterSubscriptions == 0 && registeredTextures == 0
      && eventMonitors == 0 && sessionTimers == 0 && powerAssertions == 0
      && writers == 0 && compositors == 0
  }

  /// The wire map. Every value an `Int`, `writers` and `compositors` included:
  /// one type for the whole map is what lets Dart compare it in one equality.
  public var map: [String: Any] {
    [
      "captureStreams": captureStreams,
      "cameraSessions": cameraSessions,
      "microphoneSessions": microphoneSessions,
      "meteringTaps": meteringTaps,
      "meterSubscriptions": meterSubscriptions,
      "registeredTextures": registeredTextures,
      "overlayEngines": overlayEngines,
      "eventMonitors": eventMonitors,
      "sessionTimers": sessionTimers,
      "powerAssertions": powerAssertions,
      "writers": writers,
      "compositors": compositors,
    ]
  }
}

/// The rows a live session owns, mirrored so the census can be read from a
/// thread that is not the one that built them (§19.1).
///
/// A mirror rather than a direct read of `RecordingSession`'s own fields, and
/// this is not tidiness: those fields are written on the concurrency pool and
/// the census is asked for on the platform thread, and an unsynchronised
/// read/write of a class reference can over-release rather than merely read
/// stale — the hazard `sessionLock` already exists for in the plugin. Six
/// booleans behind one lock cost nothing and cannot crash.
///
/// The camera and the microphone are deliberately *not* here: they are `let`
/// captures the session owns for its whole life, so they can be asked directly
/// and there is nothing to keep in step.
public final class SessionResourceLedger: @unchecked Sendable {
  private let lock = NSLock()
  private var captureStream = false
  private var writer = false
  private var compositor = false
  private var timer = false
  private var powerAssertion = false

  public init() {}

  public func noteCaptureStream(_ held: Bool) { write { self.captureStream = held } }
  public func noteWriter(_ held: Bool) { write { self.writer = held } }
  public func noteCompositor(_ held: Bool) { write { self.compositor = held } }
  public func noteTimer(_ held: Bool) { write { self.timer = held } }
  public func notePowerAssertion(_ held: Bool) { write { self.powerAssertion = held } }

  /// Everything the session built, dropped at once — what `tearDown` does.
  public func releaseAll() {
    write {
      self.captureStream = false
      self.writer = false
      self.compositor = false
      self.timer = false
      self.powerAssertion = false
    }
  }

  public var census: ResourceCensus {
    lock.lock()
    defer { lock.unlock() }
    return ResourceCensus(
      captureStreams: captureStream ? 1 : 0,
      sessionTimers: timer ? 1 : 0,
      powerAssertions: powerAssertion ? 1 : 0,
      writers: writer ? 1 : 0,
      compositors: compositor ? 1 : 0)
  }

  private func write(_ body: () -> Void) {
    lock.lock()
    body()
    lock.unlock()
  }
}

File: packages/recorder_platform_interface/lib/src/models/resource_census.dart
/// What the host is still holding, counted (§19.1).
///
/// §19.1 specifies what a session must have released by the time it is over,
/// and observes that "state is cleaned up" is not a requirement because nothing
/// can fail it. This is what makes it falsifiable: every row is an integer a
/// test can compare, and the two tests §19.1 names — census equality across ten
/// start → stop cycles, and every row zero after one stop — are both ordinary
/// equality assertions on this object.
///
/// **Debug-only, and diagnostic only.** Nothing in the application may branch
/// on a census: it is an assertion surface, not a source of truth about the
/// session. A host that cannot count a row reports zero for it rather than
/// failing the call, because a census that throws on the teardown path would
/// turn a leak into a crash.
///
/// Every field is a count, including [writers] and [compositors], which are 0
/// or 1. One type for the whole map is what lets equality be a single
/// comparison rather than twelve — and a writer that somehow existed twice
/// should read as 2, not as `true`.
class ResourceCensus {
  const ResourceCensus({
    this.captureStreams = 0,
    this.cameraSessions = 0,
    this.microphoneSessions = 0,
    this.meteringTaps = 0,
    this.meterSubscriptions = 0,
    this.registeredTextures = 0,
    this.overlayEngines = 0,
    this.eventMonitors = 0,
    this.sessionTimers = 0,
    this.powerAssertions = 0,
    this.writers = 0,
    this.compositors = 0,
  });

  /// Live screen-capture streams and their content filters.
  final int captureStreams;

  /// Camera capture sessions that still have a device input attached.
  ///
  /// A *stopped* session still counts while its input is attached: §19.1
  /// releases "the camera capture **and its device input**", and a stopped
  /// capture session that keeps its input is exactly the idle recorder owning a
  /// configured camera graph that `releaseSession` exists to end.
  final int cameraSessions;

  /// The microphone's equivalent, on the same terms.
  final int microphoneSessions;

  /// Metering taps the *meter* has open.
  ///
  /// Never a level read off a capture the session already holds, which opens
  /// nothing and so counts nothing: §33.7 forbids opening a device a recording
  /// is using a second time, and a census that counted that as a tap would
  /// report a leak for behaving correctly.
  final int meteringTaps;

  /// How many callers are still asking to be metered.
  ///
  /// Counted separately from [meteringTaps] because they answer different
  /// questions: a leaked subscription with no tap is a meter that will re-open
  /// a device on the next start, and a tap with no subscribers is a device held
  /// for a bar nobody is looking at.
  final int meterSubscriptions;

  /// Textures registered against any engine — the camera preview's, today.
  final int registeredTextures;

  /// Secondary Flutter engines hosting overlay windows.
  ///
  /// Expected to be non-zero on a host that keeps its engines for the life of
  /// the process, which §19.1 allows explicitly and
  /// `docs/development/compatibility-matrix.md` records per platform. What must
  /// not change is the number, which is what the equality test asserts.
  final int overlayEngines;

  /// Event monitors, hooks and observers a *session* installed — menu
  /// dismissal, drag end, low-level input hooks.
  ///
  /// Not the ones the plugin installs once for the life of the process: those
  /// are part of the launch census, and the equality test is what holds them to
  /// it.
  final int eventMonitors;

  /// Session timers: the tick, and anything else scheduled per session.
  final int sessionTimers;

  /// Power/sleep assertions held.
  final int powerAssertions;

  /// 1 while an asset writer, its inputs and its pixel-buffer pool exist.
  final int writers;

  /// 1 while a compositor and its caches exist.
  final int compositors;

  /// True when every row of §19.1's *first* table is zero — what its
  /// post-session test asserts.
  ///
  /// [overlayEngines] is deliberately excluded, and this is the whole reason
  /// the getter exists rather than an `isEmpty`. §19.1 puts the engines in its
  /// second table, "survives, because it is a setting and not state": a host
  /// that keeps its overlay engines for the life of the process is lawful, and
  /// asserting zero here would fail a platform for a choice the specification
  /// grants it. What that host owes instead is stability, which the census
  /// **equality** test is what checks.
  ///
  /// Every other row is required to be zero, [eventMonitors] included: it
  /// counts only what a *session* installed, so §19.1's "the census reports the
  /// launch counts" and zero are the same number.
  bool get sessionResourcesReleased =>
      captureStreams == 0 &&
      cameraSessions == 0 &&
      microphoneSessions == 0 &&
      meteringTaps == 0 &&
      meterSubscriptions == 0 &&
      registeredTextures == 0 &&
      eventMonitors == 0 &&
      sessionTimers == 0 &&
      powerAssertions == 0 &&
      writers == 0 &&
      compositors == 0;

  Map<String, Object?> toMap() => <String, Object?>{
    'captureStreams': captureStreams,
    'cameraSessions': cameraSessions,
    'microphoneSessions': microphoneSessions,
    'meteringTaps': meteringTaps,
    'meterSubscriptions': meterSubscriptions,
    'registeredTextures': registeredTextures,
    'overlayEngines': overlayEngines,
    'eventMonitors': eventMonitors,
    'sessionTimers': sessionTimers,
    'powerAssertions': powerAssertions,
    'writers': writers,
    'compositors': compositors,
  };

  /// A missing row decodes as zero rather than failing.
  ///
  /// The census is an assertion surface: a host that has not learned to count
  /// something yet should make the test that reads it fail on the row it cannot
  /// answer, not on the decode — and a decode that threw would take the
  /// teardown path down with it.
  static ResourceCensus fromMap(Map<String, Object?> map) {
    int at(String key) => (map[key] as num? ?? 0).toInt();
    return ResourceCensus(
      captureStreams: at('captureStreams'),
      cameraSessions: at('cameraSessions'),
      microphoneSessions: at('microphoneSessions'),
      meteringTaps: at('meteringTaps'),
      meterSubscriptions: at('meterSubscriptions'),
      registeredTextures: at('registeredTextures'),
      overlayEngines: at('overlayEngines'),
      eventMonitors: at('eventMonitors'),
      sessionTimers: at('sessionTimers'),
      powerAssertions: at('powerAssertions'),
      writers: at('writers'),
      compositors: at('compositors'),
    );
  }

  @override
  bool operator ==(Object other) =>
      other is ResourceCensus &&
      other.captureStreams == captureStreams &&
      other.cameraSessions == cameraSessions &&
      other.microphoneSessions == microphoneSessions &&
      other.meteringTaps == meteringTaps &&
      other.meterSubscriptions == meterSubscriptions &&
      other.registeredTextures == registeredTextures &&
      other.overlayEngines == overlayEngines &&
      other.eventMonitors == eventMonitors &&
      other.sessionTimers == sessionTimers &&
      other.powerAssertions == powerAssertions &&
      other.writers == writers &&
      other.compositors == compositors;

  @override
  int get hashCode => Object.hash(
    captureStreams,
    cameraSessions,
    microphoneSessions,
    meteringTaps,
    meterSubscriptions,
    registeredTextures,
    overlayEngines,
    eventMonitors,
    sessionTimers,
    powerAssertions,
    writers,
    compositors,
  );

  /// Every row, named. A census that fails an equality assertion has to say
  /// *which* row moved, or the test reports "these two objects differ".
  @override
  String toString() =>
      'ResourceCensus('
      '${toMap().entries.map((MapEntry<String, Object?> e) => '${e.key}: ${e.value}').join(', ')})';
}
